pkg/algorithm/sku_scheduler.go

```diff
@@ -0,0 +1,512 @@
+		copy(bestAffinityCellIndices, currentCellIndices)
+		for i := 0; i < len(currentCellIndices); i++ {
+			bestAffinityCells[i] = freeCells[currentCellIndices[i]]
+		}
+		if affinity == optimalAffinity {
+			return affinity, true
+		} else {
+			return affinity, false
+		}
+	}
+	return bestAffinity, false
+}
+
+func (s *skuScheduler) createSkuClusterView(
+	within *skuCell,
+	withinLevel CellLevel,
+	priority CellPriority) skuClusterView {
+
+	cv := skuClusterView{}
+	for l := withinLevel; l >= CellLevel(1); l-- {
+		for _, c := range s.chainCellList[l] {
+			if (within != nil && !isAncestor(within.cell, c)) ||
+				cv.containsAncestor(ancestorNoHigherThanLevel(withinLevel, c)) {
+				continue
+			}
+			cell := &skuCell{
+				cell:                          c,
+				freeLeafCellNumAtPriority:     c.GetTotalLeafCellNum(),
+				usedLeafCellNumAtPriority:     0,
+				usedLeafCellNumHigherPriority: 0,
+				healthy:                       true,
+				address:                       "",
+			}
+			for p, num := range c.GetUsedLeafCellNumAtPriorities() {
+				if p >= priority {
+					cell.freeLeafCellNumAtPriority -= num
+				}
+				if s.crossPriorityPack {
+					cell.usedLeafCellNumAtPriority += num
+				} else {
+					if p == priority {
+						cell.usedLeafCellNumAtPriority += num
+					}
+					if p > priority {
+						cell.usedLeafCellNumHigherPriority += num
+					}
+				}
+			}
+			switch v := c.(type) {
+			case *PhysicalCell:
+				cell.healthy = v.IsHealthy()
+				cell.address = c.GetAddress()
+			case *VirtualCell:
+				if pn := v.GetPhysicalCell(); pn != nil {
+					cell.healthy = pn.IsHealthy()
+					cell.address = pn.GetAddress()
+				}
+			}
+			cv = append(cv, cell)
+		}
+	}
+	sort.Stable(cv)
+	return cv
+}
+
+// Len method for sorting sku cells in cluster view
+func (cv skuClusterView) Len() int {
+	return len(cv)
+}
+
+// Less method for sorting sku cells in cluster view
+// sort in the following order:
+// 1. cell health (prefer healthy)
+// 2. cell level (prefer lower)
+// 3. usedLeafCellNumAtPriority (prefer higher)
+// 4. usedLeafCellNumHigherPriority (prefer lower)
+// 5. cell address (prefer lower)
+func (cv skuClusterView) Less(i, j int) bool {
+	if cv[i].healthy != cv[j].healthy {
+		return cv[i].healthy
+	}
+	if cv[i].cell.GetLevel() != cv[j].cell.GetLevel() {
+		return cv[i].cell.GetLevel() < cv[j].cell.GetLevel()
+	}
+	if cv[i].usedLeafCellNumAtPriority != cv[j].usedLeafCellNumAtPriority {
+		return cv[i].usedLeafCellNumAtPriority > cv[j].usedLeafCellNumAtPriority
+	}
+	if cv[i].usedLeafCellNumHigherPriority != cv[j].usedLeafCellNumHigherPriority {
+		return cv[i].usedLeafCellNumHigherPriority < cv[j].usedLeafCellNumHigherPriority
+	}
+	if cv[i].address != cv[j].address {
+		return cv[i].address < cv[j].address
+	}
+	return true
+}
+
+// Swap method for sorting sku cells in cluster view
+func (cv skuClusterView) Swap(i int, j int) {
+	cv[i], cv[j] = cv[j], cv[i]
+}
+
+func (cv skuClusterView) containsAncestor(cell Cell) bool {
```

Comment:
Suggest to use "contains" instead of "containsAncestor" because this function does not have the logic of "ancestor".

Reply:
updated